cmd/watt/aggregator.go

```diff
@@ -60,14 +60,97 @@ func NewAggregator(snapshots chan<- string, k8sWatches chan<- []KubernetesWatchS
+	//    from the kubernetesEventProcessor channel and performing
+	//    a blocking a.maybeNotify(). This means that we can only
+	//    *write* to the kubernetesEventProcessor channel when we are
+	//    not currently processing an event, but when that happens, we
+	//    will still read from a.KubernetesEvents and a.ConsulEvents
+	//    and update potentialKubernetesEventSignal.
+	//
+	// There are three caveats to the above:
+	//
+	// 1. At startup, we don't yet have a event to write, but
+	//    we're not processing anything, so we will try to write
+	//    something down the kubernetesEventProcessor channel.
+	//    To cope with this, the invoking goroutine will ignore events
+	//    signals that have a event.skip flag.
+	//
+	// 2. If we process an event quickly, or if there aren't new
+	//    events available, then we end up busy looping and
+	//    sending the same potentialKubernetesEventSignal value down
+	//    the kubernetesEventProcessor channel multiple times. To cope
+	//    with this, whenever we have successfully written to the
+	//    kubernetesEventProcessor channel, we do a *blocking* read of
+	//    the next event from a.KubernetesEvents and a.ConsulEvents.
+	//
+	// 3. Always be calling a.setKubernetesResources as soon as we
+	//    receive an event. This is a fast non-blocking call that
+	//    update watches, we can't coalesce this call.
+
 	p.Ready()
 
+	type eventSignal struct {
+		kubernetesEvent k8sEvent
+		skip            bool
+	}
+
+	kubernetesEventProcessor := make(chan eventSignal)
+	go func() {
+		for event := range kubernetesEventProcessor {
+			if event.skip {
+				// ignore the initial eventSignal to deal with the
+				// corner case where we haven't yet received an event yet.
+				continue
+			}
+			a.maybeNotify(p)
+		}
+	}()
+
+	potentialKubernetesEventSignal := eventSignal{kubernetesEvent: k8sEvent{}, skip: true}
 	for {
 		select {
-		case event := <-a.KubernetesEvents:
-			a.setKubernetesResources(event)
-			a.maybeNotify(p)
+		case potentialKubernetesEvent := <-a.KubernetesEvents:
+			// if a new KubernetesEvents is available to be read,
+			// and we can't write to the kubernetesEventProcessor channel,
+			// then we will overwrite potentialKubernetesEvent
+			// with a newer event while still processing a.setKubernetesResources
+			a.setKubernetesResources(potentialKubernetesEvent)
+			potentialKubernetesEventSignal = eventSignal{kubernetesEvent: potentialKubernetesEvent, skip: false}
+		case kubernetesEventProcessor <- potentialKubernetesEventSignal:
+			// if we aren't currently blocked in
+			// a.maybeNotify() then the above goroutine will be
+			// reading from the kubernetesEventProcessor channel and we
+			// will send the current potentialKubernetesEventSignal
+			// value over the kubernetesEventProcessor channel to be
+			// processed
+			select {
+			case potentialKubernetesEvent := <-a.KubernetesEvents:
+				a.setKubernetesResources(potentialKubernetesEvent)
```

Comment:
Next time you're here, a comment about "here's the blocking read I was talking about!" wouldn't be amiss. Not going to hold for that though.